Error PKG-1 when file not found

When the validator is executed, and the file is not found, an error is thrown. But the error message is:

PKG-1: alm.ods [ERROR] An OpenDocument Package shall be a well formed Zip Archive.
POL-2: alm.ods [ERROR] Standard Compliance | The file MUST comply with the standard "OASIS Open Document Document Format for Office Applications (OpenDocument) v1.3".

This is somewhat misleading when the file does in fact not exist.

Ok, so this only happens, when I run the validator with -p (but of course, I use that in most cases). I think the error produced when running it without -p is fine, it just needs to also produce the error when running with -p.

C:\odf\odf-validator-main>odf-validator.bat "filer/testfiler/xxx.ods"
APP-1: [INFO] Validating filer\testfiler\xxx.ods.
APP-2: [ERROR] Supplied Path parameter: filer\testfiler\xxx.ods  does not exist.

C:\odf\odf-validator-main>odf-validator.bat -p filer/testfiler/xxx.ods"
APP-1: [INFO] Validating filer\testfiler\xxx.ods.
APP-5: [INFO] DNA ODF Spreadsheets Preservation Specification Profile report for filer\testfiler\xxx.ods.
PKG-1: xxx.ods [ERROR] An OpenDocument Package SHALL be a well formed Zip Archive.
POL_2: xxx.ods [ERROR] Standard Compliance | The file MUST comply with the standard "OASIS Open Document Format for Office Applications (OpenDocument) v1.3".
NOT VALID, 2 errors, 0 warnings and 0 info messages.

Fix is working for this issue. I now get the same error message whether I run it with or without -p.

C:\odf\odf-validator-fix-missing-file-profile>odf-validator.bat -p testfiler/T001.ods
APP-1: [INFO] Validating testfiler\T001.ods.
APP-2: [ERROR] Supplied Path parameter: testfiler\T001.ods  does not exist.

C:\odf\odf-validator-fix-missing-file-profile>odf-validator.bat testfiler/T001.ods
APP-1: [INFO] Validating testfiler\T001.ods.
APP-2: [ERROR] Supplied Path parameter: testfiler\T001.ods  does not exist.

I can see in the changes to CliValidator that you aligned the checks for with/without profile for the FileNotFound exception. But I noticed that the IOException (which was moved to line 82) is not identical to the equivalent line for running the validator without the profile (line 71) which contains three exceptions (ParserConfigurationException | SAXException | IOException). I have not managed to trigger any of these exceptions, but I was wondering whether there is a reason for the discrepancy?

--- a/odf-apps/pom.xml
+++ b/odf-apps/pom.xml
@@ -5,7 +5,7 @@
   <parent>
     <groupId>org.openpreservation.odf</groupId>
     <artifactId>odf-validator</artifactId>
-    <version>0.10.0</version>
+    <version>0.11.0</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 

--- a/odf-apps/src/main/java/org/openpreservation/odf/apps/CliValidator.java
+++ b/odf-apps/src/main/java/org/openpreservation/odf/apps/CliValidator.java
@@ -37,7 +37,7 @@ class CliValidator implements Callable<Integer> {
     private static final MessageFactory FACTORY = Messages
             .getInstance("org.openpreservation.odf.apps.messages.Messages");
 
-    @Option(names = {"-p", "--profile"}, description = "Validate using extended Spreadsheet preservation profile.")
+    @Option(names = { "-p", "--profile" }, description = "Validate using extended Spreadsheet preservation profile.")
     private boolean profileFlag;
     @Parameters(paramLabel = "FILE", arity = "1..*", description = "A list of Open Document Format spreadsheet files to validate.")
     private File[] toValidateFiles;
@@ -77,8 +77,10 @@ class CliValidator implements Callable<Integer> {
     private ProfileResult profilePath(final Path toProfile) {
         try {
             return dnaProfile.check(parser.parsePackage(toProfile));
-        } catch (IllegalArgumentException | IOException e) {
+        } catch (IllegalArgumentException | FileNotFoundException e) {
             this.logMessage(toProfile, Messages.getMessageInstance("APP-2", Severity.ERROR, e.getMessage()));
+        } catch (IOException e) {
+            e.printStackTrace();
         }
         return null;
     }
@@ -92,7 +94,8 @@ class CliValidator implements Callable<Integer> {
         Integer retStatus = 0;
         for (Map.Entry<Path, ValidationReport> entry : validationReports.entrySet()) {
             if (this.appMessages.containsKey(entry.getKey())) {
-                retStatus = Math.max(retStatus, processMessageLists(this.appMessages.get(entry.getKey()).getMessages().values()));
+                retStatus = Math.max(retStatus,
+                        processMessageLists(this.appMessages.get(entry.getKey()).getMessages().values()));
             }
             if (entry.getValue() != null) {
                 retStatus = Math.max(retStatus, results(entry.getKey(), entry.getValue()));
@@ -100,7 +103,8 @@ class CliValidator implements Callable<Integer> {
         }
         for (Map.Entry<Path, ProfileResult> entry : profileResults.entrySet()) {
             if (this.appMessages.containsKey(entry.getKey())) {
-                retStatus = Math.max(retStatus, processMessageLists(this.appMessages.get(entry.getKey()).getMessages().values()));
+                retStatus = Math.max(retStatus,
+                        processMessageLists(this.appMessages.get(entry.getKey()).getMessages().values()));
             }
             if (entry.getValue() != null) {
                 retStatus = Math.max(retStatus, results(entry.getKey(), entry.getValue()));
@@ -159,7 +163,9 @@ class CliValidator implements Callable<Integer> {
         if (report.getValidationReport() != null && report.getValidationReport().documentMessages.hasErrors()) {
             retStatus = 1;
         }
-        MessageLog profileMessages = (report.getValidationReport() != null) ? report.getValidationReport().documentMessages : Messages.messageLogInstance();
+        MessageLog profileMessages = (report.getValidationReport() != null)
+                ? report.getValidationReport().documentMessages
+                : Messages.messageLogInstance();
         profileMessages.add(report.getProfileMessages().getMessages());
         outputSummary(profileMessages);
         return retStatus;
@@ -185,10 +191,12 @@ class CliValidator implements Callable<Integer> {
                     messageLog.getWarningCount(), messageLog.getInfoCount()));
         } else {
             ConsoleFormatter
-                    .info(String.format("VALID, no errors, no warnings and %d info message found.", messageLog.getInfoCount()));
+                    .info(String.format("VALID, no errors, no warnings and %d info message found.",
+                            messageLog.getInfoCount()));
         }
         ConsoleFormatter.newline();
     }
+
     private final void logMessage(final Path path, final Message message) {
         this.appMessages.putIfAbsent(path, Messages.messageLogInstance());
         this.appMessages.get(path).add(path.toString(), message);

--- a/odf-core/pom.xml
+++ b/odf-core/pom.xml
@@ -5,7 +5,7 @@
   <parent>
     <groupId>org.openpreservation.odf</groupId>
     <artifactId>odf-validator</artifactId>
-    <version>0.10.0</version>
+    <version>0.11.0</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 

--- a/odf-core/src/main/java/org/openpreservation/odf/pkg/PackageParserImpl.java
+++ b/odf-core/src/main/java/org/openpreservation/odf/pkg/PackageParserImpl.java
@@ -88,6 +88,7 @@ final class PackageParserImpl implements PackageParser {
     }
 
     private final OdfPackage parsePackage(final Path toParse, final String name) throws IOException {
+        Checks.existingFileCheck(toParse);
         this.initialise();
         try {
             this.format = sniff(toParse);

--- a/odf-core/src/main/java/org/openpreservation/odf/validation/Validator.java
+++ b/odf-core/src/main/java/org/openpreservation/odf/validation/Validator.java
@@ -1,7 +1,6 @@
 package org.openpreservation.odf.validation;
 
 import java.io.File;
-import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -55,6 +54,7 @@ public class Validator {
     public ValidationReport validateSingleFormat(final Path toValidate, final Formats legal) throws ParserConfigurationException, IOException, SAXException {
         Objects.requireNonNull(toValidate, String.format(Checks.NOT_NULL, "Path", TO_VAL_STRING));
         Objects.requireNonNull(legal, String.format(Checks.NOT_NULL, "Formats", "legal"));
+        Checks.existingFileCheck(toValidate);
         ValidationReport report = validate(toValidate);
         if (report.document == null || report.document.getFormat() == null) {
             report.add(toValidate.toString(), FACTORY.getError("DOC-6"));
@@ -77,7 +77,7 @@ public class Validator {
         Objects.requireNonNull(toValidate, String.format(Checks.NOT_NULL, "Path", TO_VAL_STRING));
 
         // Check if the path exists and is not a directory
-        existingFileCheck(toValidate);
+        Checks.existingFileCheck(toValidate);
 
         if (OdfPackages.isZip(toValidate)) {
             return validatePackage(toValidate);
@@ -92,14 +92,6 @@ public class Validator {
         final ValidationReport report = ValidationReport.of(toValidate.toString());
         report.add(toValidate.toString(), FACTORY.getError("DOC-1"));
         return report;
-    }
-
-    static final void existingFileCheck(final Path toValidate) throws FileNotFoundException {
-        if (!Files.exists(toValidate)) {
-            throw new FileNotFoundException(errMessage(toValidate.toString(), " does not exist."));
-        } else if (Files.isDirectory(toValidate)) {
-            throw new IllegalArgumentException(errMessage(toValidate.toString(), " is a directory."));
-        }
     }
 
     private ValidationReport validatePackage(final Path toValidate) throws ParserConfigurationException, SAXException, IOException {
@@ -132,9 +124,5 @@ public class Validator {
         }
         report.add(toValidate.toString(), parseResult.getMessages());
         return report;
-    }
-
-    private static final String errMessage(final String toValidate, final String subMess) {
-        return String.format("Supplied Path parameter: %s %s", toValidate, subMess);
     }
 }

--- a/odf-core/src/main/java/org/openpreservation/utils/Checks.java
+++ b/odf-core/src/main/java/org/openpreservation/utils/Checks.java
@@ -1,5 +1,9 @@
 package org.openpreservation.utils;
 
+import java.io.FileNotFoundException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
 public class Checks {
     public static final String NOT_NULL = "%s parameter: %s must not be null.";
     public static final String NOT_EMPTY = "String parameter: %s must not be empty.";
@@ -7,5 +11,17 @@ public class Checks {
 
     private Checks() {
         throw new AssertionError("Should not be instantiated.");
+    }
+
+    public static final void existingFileCheck(final Path toValidate) throws FileNotFoundException {
+        if (!Files.exists(toValidate)) {
+            throw new FileNotFoundException(errMessage(toValidate.toString(), " does not exist."));
+        } else if (Files.isDirectory(toValidate)) {
+            throw new IllegalArgumentException(errMessage(toValidate.toString(), " is a directory."));
+        }
+    }
+
+    public static final String errMessage(final String toValidate, final String subMess) {
+        return String.format("Supplied Path parameter: %s %s", toValidate, subMess);
     }
 }

--- a/odf-core/src/test/java/org/openpreservation/odf/pkg/PackageParserTest.java
+++ b/odf-core/src/test/java/org/openpreservation/odf/pkg/PackageParserTest.java
@@ -7,6 +7,7 @@ import static org.junit.Assert.assertThrows;
 import static org.junit.Assert.assertTrue;
 
 import java.io.File;
+import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.InputStream;
 import java.net.URISyntaxException;
@@ -39,11 +40,55 @@ public class PackageParserTest {
     }
 
     @Test
+    public void testParseMissingFile() throws ParserConfigurationException, SAXException {
+        PackageParser parser = OdfPackages.getPackageParser();
+        File file = new File("missingFileRubbishdhbna");
+        assertThrows("FileNotFoundException expected",
+                FileNotFoundException.class,
+                () -> {
+                    parser.parsePackage(file);
+                });
+    }
+
+    @Test
+    public void testParseDirFile() throws ParserConfigurationException, SAXException {
+        PackageParser parser = OdfPackages.getPackageParser();
+        File file = new File(".");
+        assertThrows("IllegalArgumentException expected",
+                IllegalArgumentException.class,
+                () -> {
+                    parser.parsePackage(file);
+                });
+    }
+
+    @Test
     public void testParseNullPath() throws ParserConfigurationException, SAXException {
         PackageParser parser = OdfPackages.getPackageParser();
         Path path = null;
         assertThrows("NullPointerException expected",
                 NullPointerException.class,
+                () -> {
+                    parser.parsePackage(path);
+                });
+    }
+
+    @Test
+    public void testParseMissingPath() throws ParserConfigurationException, SAXException {
+        PackageParser parser = OdfPackages.getPackageParser();
+        Path path = Path.of(".", "missingFileRubbishdhbna");
+        assertThrows("FileNotFoundException expected",
+                FileNotFoundException.class,
+                () -> {
+                    parser.parsePackage(path);
+                });
+    }
+
+    @Test
+    public void testParseDirPath() throws ParserConfigurationException, SAXException {
+        PackageParser parser = OdfPackages.getPackageParser();
+        Path path = Path.of(".");
+        assertThrows("IllegalArgumentException expected",
+                IllegalArgumentException.class,
                 () -> {
                     parser.parsePackage(path);
                 });
@@ -66,30 +111,36 @@ public class PackageParserTest {
     }
 
     @Test
-    public void testParsePackagePath() throws ParserConfigurationException, SAXException, URISyntaxException, IOException {
+    public void testParsePackagePath()
+            throws ParserConfigurationException, SAXException, URISyntaxException, IOException {
         PackageParser parser = OdfPackages.getPackageParser();
         OdfPackage pkg = parser.parsePackage(Paths.get(new File(TestFiles.EMPTY_ODS.toURI()).getAbsolutePath()));
         assertNotNull("Parsed package should not be null", pkg);
         assertTrue("Package should have a mimetype entry", pkg.hasMimeEntry());
-        assertEquals("Mimetype should be Spreadsheet", "application/vnd.oasis.opendocument.spreadsheet", pkg.getMimeType());
+        assertEquals("Mimetype should be Spreadsheet", "application/vnd.oasis.opendocument.spreadsheet",
+                pkg.getMimeType());
     }
 
     @Test
-    public void testParsePackageFile() throws ParserConfigurationException, SAXException, URISyntaxException, IOException {
+    public void testParsePackageFile()
+            throws ParserConfigurationException, SAXException, URISyntaxException, IOException {
         PackageParser parser = OdfPackages.getPackageParser();
         OdfPackage pkg = parser.parsePackage(new File(TestFiles.EMPTY_ODS.toURI()));
         assertNotNull("Parsed package should not be null", pkg);
         assertTrue("Package should have a mimetype entry", pkg.hasMimeEntry());
-        assertEquals("Mimetype should be Spreadsheet", "application/vnd.oasis.opendocument.spreadsheet", pkg.getMimeType());
+        assertEquals("Mimetype should be Spreadsheet", "application/vnd.oasis.opendocument.spreadsheet",
+                pkg.getMimeType());
     }
 
     @Test
-    public void testParsePackageStream() throws ParserConfigurationException, SAXException, URISyntaxException, IOException {
+    public void testParsePackageStream()
+            throws ParserConfigurationException, SAXException, URISyntaxException, IOException {
         PackageParser parser = OdfPackages.getPackageParser();
         OdfPackage pkg = parser.parsePackage(TestFiles.EMPTY_ODS.openStream(), TestFiles.EMPTY_ODS.toString());
         assertNotNull("Parsed package should not be null", pkg);
         assertTrue("Package should have a mimetype entry", pkg.hasMimeEntry());
-        assertEquals("Mimetype should be Spreadsheet", "application/vnd.oasis.opendocument.spreadsheet", pkg.getMimeType());
+        assertEquals("Mimetype should be Spreadsheet", "application/vnd.oasis.opendocument.spreadsheet",
+                pkg.getMimeType());
     }
 
     @Test
@@ -97,16 +148,17 @@ public class PackageParserTest {
         PackageParser parser = OdfPackages.getPackageParser();
         OdfPackage pkg = parser.parsePackage(TestFiles.DSIG_INVALID.openStream(), TestFiles.DSIG_INVALID.toString());
         ParseResult result = pkg.getEntryXmlParseResult("META-INF/documentsignatures.xml");
-        assertNotNull("Dsig file META-INF/documentsignatures.xml result should not be null" , result);
-        assertTrue("Package should have a well formed dsig for META-INF/documentsignatures.xml" , result.isWellFormed());
+        assertNotNull("Dsig file META-INF/documentsignatures.xml result should not be null", result);
+        assertTrue("Package should have a well formed dsig for META-INF/documentsignatures.xml", result.isWellFormed());
     }
 
     @Test
     public void testManifestNotWF() throws IOException {
         PackageParser parser = OdfPackages.getPackageParser();
-        OdfPackage pkg = parser.parsePackage(TestFiles.MANIFEST_NOT_WF.openStream(), TestFiles.MANIFEST_NOT_WF.toString());
+        OdfPackage pkg = parser.parsePackage(TestFiles.MANIFEST_NOT_WF.openStream(),
+                TestFiles.MANIFEST_NOT_WF.toString());
         ParseResult result = pkg.getEntryXmlParseResult("META-INF/manifest.xml");
-        assertNotNull("Dsig file META-INF/documentsignatures.xml result should not be null" , result);
-        assertFalse("Package should NOT have a well formed META-INF/manifest.xml" , result.isWellFormed());
+        assertNotNull("Dsig file META-INF/documentsignatures.xml result should not be null", result);
+        assertFalse("Package should NOT have a well formed META-INF/manifest.xml", result.isWellFormed());
     }
 }

--- a/odf-reporting/pom.xml
+++ b/odf-reporting/pom.xml
@@ -5,7 +5,7 @@
   <parent>
     <groupId>org.openpreservation.odf</groupId>
     <artifactId>odf-validator</artifactId>
-    <version>0.10.0</version>
+    <version>0.11.0</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 

--- a/odf-validator
+++ b/odf-validator
@@ -78,5 +78,5 @@ fi
 
 exec "$JAVACMD" -Dfile.encoding=UTF8 -XX:+IgnoreUnrecognizedVMOptions \
   -Dapp.name="ODF Validator" \
-  -jar odf-apps/target/odf-apps-0.10.0-jar-with-dependencies.jar \
+  -jar odf-apps/target/odf-apps-0.11.0-jar-with-dependencies.jar \
   "$@"

--- a/odf-validator.bat
+++ b/odf-validator.bat
@@ -85,7 +85,7 @@ if NOT "%CLASSPATH_PREFIX%" == "" set CLASSPATH=%CLASSPATH_PREFIX%;%CLASSPATH%
 @REM Reaching here means variables are defined and arguments have been captured
 :endInit
 
-"%JAVACMD%" -Dfile.encoding=UTF8 -XX:+IgnoreUnrecognizedVMOptions -Dapp.name="ODF Validator" -jar .\odf-apps\target\odf-apps-0.10.0-jar-with-dependencies.jar %CMD_LINE_ARGS%
+"%JAVACMD%" -Dfile.encoding=UTF8 -XX:+IgnoreUnrecognizedVMOptions -Dapp.name="ODF Validator" -jar .\odf-apps\target\odf-apps-0.11.0-jar-with-dependencies.jar %CMD_LINE_ARGS%
 if %ERRORLEVEL% NEQ 0 goto error
 goto end
 

--- a/pom.xml
+++ b/pom.xml
@@ -10,7 +10,7 @@
 
   <groupId>org.openpreservation.odf</groupId>
   <artifactId>odf-validator</artifactId>
-  <version>0.10.0</version>
+  <version>0.11.0</version>
   <packaging>pom</packaging>
 
   <modules>

